src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -86,24 +88,55 @@
     (jruby-internal/create-pool-instance! pool new-id config
                                           (partial send-flush-instance! pool-context))))
 
-(schema/defn borrow-all-jrubies :- [JRubyInstance]
-  "Locks the pool and borrows all the instances"
-  [pool-context :- jruby-schemas/PoolContext]
+(schema/defn borrow-all-jrubies*
+  "The core logic for borrow-all-jrubies. Should not be called from borrow-all-jrubies"
+  [pool-context :- jruby-schemas/PoolContext
+   borrow-exception :- IDeref]
   (let [pool-size (jruby-internal/get-pool-size pool-context)
         pool (jruby-internal/get-pool pool-context)
         borrow-fn (partial jruby-internal/borrow-from-pool pool-context)]
-    (.lock pool)
     (try
       (into [] (repeatedly pool-size borrow-fn))
+
+      ; We catch the exception here, place it in the borrow-exception atom
+      ; for use in the calling fn, and then throw it again so that
+      ; shutdown-on-error will also catch it and shutdown the app
       (catch Exception e
         (.clear pool)
         (jruby-internal/insert-poison-pill pool e)
-        (throw (IllegalStateException.
-                (i18n/tru "There was a problem borrowing a JRubyInstance from the pool.")
-                e)))
+        (let [exception (IllegalStateException.
+                         (i18n/tru "There was a problem borrowing a JRubyInstance from the pool.")
+                         e)]
+          (reset! borrow-exception exception)
+          (throw exception)))
       (finally
         (.unlock pool)))))
 
+(schema/defn borrow-all-jrubies :- [JRubyInstance]
+  "Locks the pool and borrows all the instances"
+  [pool-context :- jruby-schemas/PoolContext]
+  (let [pool (jruby-internal/get-pool pool-context)
+        flush-timeout (jruby-internal/get-flush-timeout pool-context)
+        shutdown-on-error (get-shutdown-on-error-fn pool-context)
+        borrow-exception (atom nil)]
+    ; If lock fails, abort and throw an exception
+    (try
+      (.lockWithTimeout pool flush-timeout TimeUnit/MILLISECONDS)
+      (catch TimeoutException e
+        (sling/throw+ {:kind ::jruby-lock-timeout
+                       :msg (.getMessage e)})))
```

Comment:
This expands to `:puppetlabs.services.jruby-pool-manager.impl.jruby-agents/jruby-lock-timeout` which puppetserver or any other consumer has to reference. I'm not sure what the best way to handle this is. Maybe just explicitly throw a `:puppetlabs.services.jruby-pool-manager.jruby-core/jruby-lock-timeout` or something?